src/transport/MessageEndpoint.cpp

```diff
@@ -0,0 +1,303 @@
+  , id(faabric::util::generateGid())
+{}
+
+MessageEndpoint::~MessageEndpoint()
+{
+    if (this->socket != nullptr) {
+        faabric::util::getLogger()->warn(
+          "Destroying an open message endpoint!");
+        this->close(false);
+    }
+}
+
+void MessageEndpoint::open(faabric::transport::MessageContext& context,
+                           faabric::transport::SocketType sockType,
+                           bool bind)
+{
+    // Check we are opening from the same thread. We assert not to incur in
+    // costly checks when running a Release build.
+    assert(tid == std::this_thread::get_id());
+
+    const auto& logger = faabric::util::getLogger();
+
+    std::string address =
+      "tcp://" + this->host + ":" + std::to_string(this->port);
+
+    // Note - only one socket may bind, but several can connect. This
+    // allows for easy N - 1 or 1 - N PUSH/PULL patterns. Order between
+    // bind and connect does not matter.
+    switch (sockType) {
+        case faabric::transport::SocketType::PUSH:
+            try {
+                this->socket = std::make_unique<zmq::socket_t>(
+                  context.get(), zmq::socket_type::push);
+            } catch (zmq::error_t& e) {
+                logger->error(
+                  "Error opening SEND socket to {}: {}", address, e.what());
+                throw;
+            }
+            break;
+        case faabric::transport::SocketType::PULL:
+            try {
+                this->socket = std::make_unique<zmq::socket_t>(
+                  context.get(), zmq::socket_type::pull);
+            } catch (zmq::error_t& e) {
+                logger->error("Error opening RECV socket bound to {}: {}",
+                              address,
+                              e.what());
+                throw;
+            }
+            break;
+        default:
+            throw std::runtime_error("Unrecognized socket type");
+    }
+    assert(this->socket != nullptr);
+
+    // Bind or connect the socket
+    if (bind) {
+        try {
+            this->socket->bind(address);
+        } catch (zmq::error_t& e) {
+            logger->error("Error binding socket to {}: {}", address, e.what());
+            throw;
+        }
+    } else {
+        try {
+            this->socket->connect(address);
+        } catch (zmq::error_t& e) {
+            logger->error(
+              "Error connecting socket to {}: {}", address, e.what());
+            throw;
+        }
+    }
+}
+
+void MessageEndpoint::send(uint8_t* serialisedMsg, size_t msgSize, bool more)
+{
+    assert(tid == std::this_thread::get_id());
+    assert(this->socket != nullptr);
+
+    const auto& logger = faabric::util::getLogger();
+
+    if (more) {
+        try {
+            auto res = this->socket->send(zmq::buffer(serialisedMsg, msgSize),
+                                          zmq::send_flags::sndmore);
+            if (res != msgSize) {
+                throw std::runtime_error(
```

Comment:
I had problems _catching_ exceptions and then re-throwing `std::runtime_error`s. My takeaway is that throwing _new_ exceptions from within a `catch` block can be risky, but otherwise should be fine.

Instead of throwing new ones, I've decided to log and re-throw.

Reply:
I generally go for a detailed logging statement then a generic "why" in the exception message. Here and on line 112 you've got a detailed "why" message and no log line (i.e. using `fmt::format` in the exception constructor). In threaded code the exception message can get lost or be nested in a stack trace, whereas if we always log before throwing an exception, you can scroll up to the line just before it to see what happened.